### PR TITLE
Simplify trap handling

### PR DESCRIPTION
**Pull request by axic**
Pulled out of #219.

### CODE REVIEW

**Comment by chfast** [comment]
```
Comparing master to traponly
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     +0.0189         +0.0189            87            88            87            88
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    +0.0245         +0.0245          1307          1339          1307          1339
fizzy/execute/ecpairing/onepoint_mean                             +0.0174         +0.0174        435697        443285        435698        443288
fizzy/execute/memset/256_bytes_mean                               +0.0562         +0.0562             7             8             7             8
fizzy/execute/memset/60000_bytes_mean                             +0.0514         +0.0514          1564          1644          1564          1644
fizzy/execute/mul256_opt0/input0_mean                             +0.0089         +0.0089            27            27            27            27
fizzy/execute/mul256_opt0/input1_mean                             +0.0114         +0.0114            27            27            27            27
fizzy/execute/sha1/512_bytes_rounds_1_mean                        +0.0349         +0.0349            93            96            93            96
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0393         +0.0393          1288          1339          1288          1339
fizzy/execute/sha256/512_bytes_rounds_1_mean                      +0.0691         +0.0691            95           101            95           101
fizzy/execute/sha256/512_bytes_rounds_16_mean                     +0.0724         +0.0724          1299          1393          1299          1393
fizzy/execute/micro/eli_interpreter/halt_mean                     +0.0662         +0.0662             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  -0.0074         -0.0074             5             5             5             5
fizzy/execute/micro/host_adler32/1_mean                           +0.0317         +0.0317             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         -0.0528         -0.0528             5             4             5             4
fizzy/execute/micro/host_adler32/1000_mean                        -0.0616         -0.0616            45            43            45            43
fizzy/execute/micro/spinner/1_mean                                +0.0738         +0.0738             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             +0.0391         +0.0390            10            11            10            11
```

**Comment by axic** [comment]
Well then let's do #219. It is a bit worrying how brittle the execution loop is to changes.

**Comment by axic** [comment]
What are the number after #219 was merged?

**Comment by chfast** [comment]
```
Comparing master to trap
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     +0.0042         +0.0042            86            86            86            86
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    +0.0011         +0.0011          1306          1307          1306          1307
fizzy/execute/ecpairing/onepoint_mean                             +0.0337         +0.0337        409714        423541        409717        423545
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   +0.0622         +0.0622            98           104            98           104
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0605         +0.0605          1439          1526          1439          1526
fizzy/execute/memset/256_bytes_mean                               +0.0010         +0.0010             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             -0.0011         -0.0011          1547          1546          1547          1546
fizzy/execute/mul256_opt0/input0_mean                             -0.0130         -0.0130            27            26            27            26
fizzy/execute/mul256_opt0/input1_mean                             +0.0019         +0.0019            27            27            27            27
fizzy/execute/sha1/512_bytes_rounds_1_mean                        +0.0011         +0.0011            90            90            90            90
fizzy/execute/sha1/512_bytes_rounds_16_mean                       -0.0004         -0.0004          1256          1256          1256          1256
fizzy/execute/sha256/512_bytes_rounds_1_mean                      -0.0298         -0.0298            94            91            94            91
fizzy/execute/sha256/512_bytes_rounds_16_mean                     -0.0335         -0.0335          1293          1250          1293          1250
fizzy/execute/micro/eli_interpreter/halt_mean                     +0.0106         +0.0106             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0038         +0.0038             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             +0.0048         +0.0048             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             +0.0071         +0.0071             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0102         +0.0102          7406          7481          7406          7481
fizzy/execute/micro/host_adler32/1_mean                           +0.0052         +0.0052             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         +0.0028         +0.0028             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        +0.0019         +0.0019            29            30            29            30
fizzy/execute/micro/spinner/1_mean                                +0.0006         +0.0006             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             +0.0004         +0.0004            10            10            10            10
```

**Comment by chfast** [comment]
v2:

```
Comparing master to trap2
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0396         -0.0396            86            82            86            82
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0476         -0.0476          1306          1244          1306          1244
fizzy/execute/ecpairing/onepoint_mean                             +0.0535         +0.0535        409714        431627        409717        431630
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   +0.0352         +0.0352            98           102            98           102
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0326         +0.0326          1439          1486          1439          1486
fizzy/execute/memset/256_bytes_mean                               -0.0007         -0.0007             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             -0.0009         -0.0009          1547          1546          1547          1546
fizzy/execute/mul256_opt0/input0_mean                             -0.0108         -0.0108            27            26            27            26
fizzy/execute/mul256_opt0/input1_mean                             -0.0119         -0.0119            27            26            27            26
fizzy/execute/sha1/512_bytes_rounds_1_mean                        -0.0023         -0.0023            90            90            90            90
fizzy/execute/sha1/512_bytes_rounds_16_mean                       -0.0030         -0.0030          1256          1253          1256          1253
fizzy/execute/sha256/512_bytes_rounds_1_mean                      +0.0002         +0.0002            94            94            94            94
fizzy/execute/sha256/512_bytes_rounds_16_mean                     -0.0024         -0.0024          1293          1290          1293          1290
fizzy/execute/micro/eli_interpreter/halt_mean                     +0.0492         +0.0492             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0168         +0.0168             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             +0.0103         +0.0103             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             -0.0009         -0.0009             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0263         +0.0263          7406          7600          7406          7600
fizzy/execute/micro/host_adler32/1_mean                           -0.0025         -0.0025             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         +0.0070         +0.0070             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        +0.0079         +0.0079            29            30            29            30
fizzy/execute/micro/spinner/1_mean                                +0.0029         +0.0029             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             +0.0008         +0.0008            10            10            10            10
```

**Comment by chfast** [comment]
I re-benchmarked master.

### v1
```
Comparing master2 to trap
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0003         -0.0003            86            86            86            86
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0010         -0.0010          1308          1307          1308          1307
fizzy/execute/ecpairing/onepoint_mean                             +0.0039         +0.0039        421901        423541        421904        423545
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   +0.0156         +0.0156           103           104           103           104
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0131         +0.0131          1507          1526          1507          1526
fizzy/execute/memset/256_bytes_mean                               -0.0015         -0.0015             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             -0.0010         -0.0010          1547          1546          1547          1546
fizzy/execute/mul256_opt0/input0_mean                             -0.0301         -0.0301            27            26            27            26
fizzy/execute/mul256_opt0/input1_mean                             -0.0128         -0.0128            27            27            27            27
fizzy/execute/sha1/512_bytes_rounds_1_mean                        -0.0005         -0.0005            90            90            90            90
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0006         +0.0006          1255          1256          1255          1256
fizzy/execute/sha256/512_bytes_rounds_1_mean                      -0.0030         -0.0030            91            91            91            91
fizzy/execute/sha256/512_bytes_rounds_16_mean                     +0.0013         +0.0013          1248          1250          1248          1250
fizzy/execute/micro/eli_interpreter/halt_mean                     +0.0097         +0.0097             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  -0.0105         -0.0105             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             +0.0008         +0.0008             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             +0.0030         +0.0030             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0096         +0.0096          7410          7481          7410          7481
fizzy/execute/micro/host_adler32/1_mean                           -0.0086         -0.0086             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         -0.0144         -0.0144             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0174         -0.0174            30            30            30            30
fizzy/execute/micro/spinner/1_mean                                +0.0006         +0.0006             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             -0.0007         -0.0007            10            10            10            10
```

### v1 to v2
```
Comparing trap to trap2
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0437         -0.0437            86            82            86            82
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0486         -0.0486          1307          1244          1307          1244
fizzy/execute/ecpairing/onepoint_mean                             +0.0191         +0.0191        423541        431627        423545        431630
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   -0.0254         -0.0254           104           102           104           102
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  -0.0264         -0.0264          1526          1486          1526          1486
fizzy/execute/memset/256_bytes_mean                               -0.0017         -0.0017             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             +0.0002         +0.0002          1546          1546          1546          1546
fizzy/execute/mul256_opt0/input0_mean                             +0.0022         +0.0022            26            26            26            26
fizzy/execute/mul256_opt0/input1_mean                             -0.0138         -0.0138            27            26            27            26
fizzy/execute/sha1/512_bytes_rounds_1_mean                        -0.0034         -0.0034            90            90            90            90
fizzy/execute/sha1/512_bytes_rounds_16_mean                       -0.0026         -0.0026          1256          1253          1256          1253
fizzy/execute/sha256/512_bytes_rounds_1_mean                      +0.0309         +0.0309            91            94            91            94
fizzy/execute/sha256/512_bytes_rounds_16_mean                     +0.0322         +0.0322          1250          1290          1250          1290
fizzy/execute/micro/eli_interpreter/halt_mean                     +0.0382         +0.0382             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0129         +0.0129             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             +0.0055         +0.0055             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             -0.0079         -0.0079             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0159         +0.0159          7481          7600          7481          7600
fizzy/execute/micro/host_adler32/1_mean                           -0.0077         -0.0077             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         +0.0041         +0.0041             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        +0.0060         +0.0060            30            30            30            30
fizzy/execute/micro/spinner/1_mean                                +0.0023         +0.0023             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             +0.0005         +0.0005            10            10            10            10
```

### v1 to v3
```
Comparing trap to trap3
Benchmark                                                            Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------------
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0338         -0.0338            86            83            86            83
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0384         -0.0384          1307          1257          1307          1257
fizzy/execute/ecpairing/onepoint_mean                             +0.0003         +0.0003        423541        423678        423545        423682
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   -0.0104         -0.0104           104           103           104           103
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0021         +0.0021          1526          1530          1526          1530
fizzy/execute/memset/256_bytes_mean                               +0.0060         +0.0060             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             +0.0093         +0.0093          1546          1560          1546          1560
fizzy/execute/mul256_opt0/input0_mean                             +0.0347         +0.0347            26            27            26            27
fizzy/execute/mul256_opt0/input1_mean                             +0.0231         +0.0231            27            27            27            27
fizzy/execute/sha1/512_bytes_rounds_1_mean                        +0.0056         +0.0056            90            91            90            91
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0078         +0.0078          1256          1266          1256          1266
fizzy/execute/sha256/512_bytes_rounds_1_mean                      -0.0263         -0.0263            91            88            91            88
fizzy/execute/sha256/512_bytes_rounds_16_mean                     -0.0247         -0.0247          1250          1219          1250          1219
fizzy/execute/micro/eli_interpreter/halt_mean                     -0.0228         -0.0228             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0135         +0.0135             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             -0.0212         -0.0212             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             -0.0187         -0.0187             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0036         +0.0036          7481          7508          7481          7508
fizzy/execute/micro/host_adler32/1_mean                           +0.0023         +0.0023             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         -0.0022         -0.0022             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0028         -0.0028            30            29            30            29
fizzy/execute/micro/spinner/1_mean                                +0.0134         +0.0134             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             +0.0176         +0.0176            10            10            10            10
```

I would go with original version 1.

**Comment by axic** [comment]
@chfast are you interested remeasuring this?

**Comment by chfast** [comment]
For the GCC10, LTO:

commit 1:
```
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     +0.0119         +0.0119            83            84            83            84
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    +0.0119         +0.0119          1258          1273          1258          1273
fizzy/execute/ecpairing/onepoint_mean                             +0.0031         +0.0031        419930        421227        419934        421232
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   +0.0287         +0.0287            98           101            98           101
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0267         +0.0267          1427          1466          1428          1466
fizzy/execute/memset/256_bytes_mean                               +0.0031         +0.0031             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             +0.0002         +0.0002          1559          1560          1560          1560
fizzy/execute/mul256_opt0/input0_mean                             -0.0137         -0.0137            27            26            27            26
fizzy/execute/mul256_opt0/input1_mean                             -0.0138         -0.0138            27            27            27            27
fizzy/execute/ramanujan_pi/33_runs_mean                           +0.0175         +0.0175           132           134           132           134
fizzy/execute/sha1/512_bytes_rounds_1_mean                        +0.0027         +0.0027            91            91            91            91
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0020         +0.0020          1261          1263          1261          1263
fizzy/execute/sha256/512_bytes_rounds_1_mean                      +0.0101         +0.0101            92            93            92            93
fizzy/execute/sha256/512_bytes_rounds_16_mean                     +0.0091         +0.0091          1267          1279          1267          1279
fizzy/execute/taylor_pi/pi_1000000_runs_mean                      -0.0005         -0.0005         41398         41378         41398         41379
fizzy/execute/micro/eli_interpreter/halt_mean                     -0.0074         -0.0074             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  +0.0210         +0.0210             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             +0.0031         +0.0031             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             +0.0027         +0.0027             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             +0.0017         +0.0017          7531          7544          7531          7544
fizzy/execute/micro/host_adler32/1_mean                           -0.0128         -0.0128             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         -0.0025         -0.0025             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0045         -0.0045            30            30            30            30
fizzy/execute/micro/spinner/1_mean                                -0.0109         -0.0109             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             -0.0036         -0.0036            11            11            11            11
```

commit last (marks trap as unlikely)
```
fizzy/execute/blake2b/512_bytes_rounds_1_mean                     -0.0072         -0.0072            83            83            83            83
fizzy/execute/blake2b/512_bytes_rounds_16_mean                    -0.0018         -0.0018          1258          1256          1258          1256
fizzy/execute/ecpairing/onepoint_mean                             -0.0154         -0.0154        419930        413447        419934        413451
fizzy/execute/keccak256/512_bytes_rounds_1_mean                   +0.0444         +0.0444            98           103            98           103
fizzy/execute/keccak256/512_bytes_rounds_16_mean                  +0.0569         +0.0569          1427          1509          1428          1509
fizzy/execute/memset/256_bytes_mean                               -0.0205         -0.0205             7             7             7             7
fizzy/execute/memset/60000_bytes_mean                             -0.0249         -0.0249          1559          1521          1560          1521
fizzy/execute/mul256_opt0/input0_mean                             -0.0424         -0.0424            27            26            27            26
fizzy/execute/mul256_opt0/input1_mean                             -0.0490         -0.0490            27            26            27            26
fizzy/execute/ramanujan_pi/33_runs_mean                           +0.0204         +0.0204           132           134           132           134
fizzy/execute/sha1/512_bytes_rounds_1_mean                        +0.0143         +0.0143            91            92            91            92
fizzy/execute/sha1/512_bytes_rounds_16_mean                       +0.0153         +0.0153          1261          1280          1261          1280
fizzy/execute/sha256/512_bytes_rounds_1_mean                      -0.0083         -0.0083            92            91            92            91
fizzy/execute/sha256/512_bytes_rounds_16_mean                     -0.0087         -0.0087          1267          1256          1267          1256
fizzy/execute/taylor_pi/pi_1000000_runs_mean                      -0.0316         -0.0316         41398         40089         41398         40090
fizzy/execute/micro/eli_interpreter/halt_mean                     -0.0234         -0.0234             0             0             0             0
fizzy/execute/micro/eli_interpreter/exec105_mean                  -0.0160         -0.0160             5             5             5             5
fizzy/execute/micro/factorial/10_mean                             -0.0007         -0.0007             0             0             0             0
fizzy/execute/micro/factorial/20_mean                             -0.0112         -0.0112             1             1             1             1
fizzy/execute/micro/fibonacci/24_mean                             -0.0047         -0.0047          7531          7495          7531          7496
fizzy/execute/micro/host_adler32/1_mean                           -0.0275         -0.0275             0             0             0             0
fizzy/execute/micro/host_adler32/100_mean                         -0.0483         -0.0483             3             3             3             3
fizzy/execute/micro/host_adler32/1000_mean                        -0.0489         -0.0490            30            29            30            29
fizzy/execute/micro/spinner/1_mean                                -0.0533         -0.0533             0             0             0             0
fizzy/execute/micro/spinner/1000_mean                             -0.0707         -0.0707            11            10            11            10
```

So I think just using the first commit is good enough. Using `[[unlikely]]` should be left for later because there are more or less portable ways of implementing it. And it may not be needed with PGO.

**Review by chfast — APPROVED** [review]

File: lib/fizzy/execute.cpp [hunk 1]
```diff
@@ -1286,11 +1184,8 @@ ExecutionResult execute(
         {
             auto const rhs = stack.top().i64;
             if (rhs == 0)
-            {
-                trap = true;
-                goto end;
-            }
-            binary_op(stack, rem<uint64_t>);
+                goto trap;
+            binary_op(stack, std::modulus<uint64_t>());
```
Comment (axic) [inline lib/fizzy/execute.cpp:1188]: @chfast is this intended?

Comment (chfast) [reply lib/fizzy/execute.cpp]: No.

**Pull request merged**